spec/util/e2e-test.ts

```diff
@@ -0,0 +1,256 @@
+import { IrcServer } from "../../src/irc/IrcServer";
+import { mkdtemp, rm } from "node:fs/promises";
+import dns from 'node:dns';
+import { MatrixClient } from "matrix-bot-sdk";
+
+// Needed to make tests work on GitHub actions. Node 17+ defaults
+// to IPv6, and the homerunner domain resolves to IPv6, but the
+// runtime doesn't actually support IPv6 🤦
+dns.setDefaultResultOrder('ipv4first');
+
+
+const DEFAULT_E2E_TIMEOUT = parseInt(process.env.IRC_TEST_TIMEOUT ?? '90000', 10);
+const WAIT_EVENT_TIMEOUT = 10000;
+
+// TODO: Expose these
+const DEFAULT_PORT = parseInt(process.env.IRC_TEST_PORT ?? '6667', 10);
+const DEFAULT_ADDRESS = process.env.IRC_TEST_ADDRESS ?? "127.0.0.1";
+
+interface Opts {
+    matrixLocalparts?: string[];
+    clients?: string[];
+    timeout?: number;
+}
+
+interface DescribeEnv {
+    homeserver: ComplementHomeServer;
+    ircBridge: IrcBridge;
+    clients: TestClient[];
+}
+
+export class IrcBridgeE2ETest extends IrcServerTest {
+
+    /**
+     * Test wrapper that automatically provisions an IRC server and Matrix server
+     * @param name The test name
+     * @param fn The inner function
+     * @returns A jest describe function.
+     */
+    static describeTest(name: string, fn: (env: () => DescribeEnv) => void, opts?: Opts) {
+        return describe(name, () => {
+            jest.setTimeout(opts?.timeout ?? DEFAULT_E2E_TIMEOUT);
+            let env: IrcBridgeE2ETest;
+            beforeEach(async () => {
+                env = new IrcBridgeE2ETest();
+                await env.setUp(opts?.clients, opts?.matrixLocalparts);
+            });
+            afterEach(async () => {
+                await env.tearDown();
+            });
+            fn(() => {
+                if (!env.homeserver) {
+                    throw Error('Homeserver not defined');
+                }
+                if (!env.ircBridge) {
+                    throw Error('ircBridge not defined');
+                }
+                return { homeserver: env.homeserver, ircBridge: env.ircBridge, clients: env.clients}
+            });
+        });
+    }
+
+    public homeserver?: ComplementHomeServer;
+    public ircBridge?: IrcBridge;
+    /**
+     * TODO: Postgresql
+     */
+    private dbPath?: string;
+
+    public async setUp(clients?: string[], matrixLocalparts?: string[]): Promise<void> {
+        // Set up Matrix homeserver - Need to register the bridge bot.
+        console.log('Setting up homeserver...');
+        this.homeserver = await createHS(["ircbridge_bot", ...matrixLocalparts || []]);
+        // Set up IRC server
+        console.log('Setting up IRC clients...');
+        await super.setUp(clients);
+        // Set up an IRC bridge.
+        this.dbPath = await mkdtemp('ircbridge-');
+
+        this.ircBridge = new IrcBridge({
+            homeserver: {
+                domain: this.homeserver.domain,
+                url: this.homeserver.url,
+                bindHostname: "0.0.0.0",
```

Comment:
Don't we expect these to always run on the same machine?
```suggestion
                bindHostname: "127.0.0.1",
```